File: clang/ads/binary_tree.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
  int data;
  struct Node* left;
  struct Node* right;
} Node;

Node* create_node(int data);
Node* search_node(Node *root, int data);
Node* get_deepest_right_node(Node *root);
void  inorder_traversal(Node *root);
void  insert_node(Node **root, int data);
void  delete_node(Node **root, int data);
void  del_deepest_right_node(Node *root, Node *d_node);

Node *create_node(int data) {
  Node *node = malloc(sizeof(Node));
  node->data = data;
  node->left = NULL;
  node->right = NULL;
  return node;
}

void insert_node(Node **root, int data) {
  Node *node = create_node(data);
  if (*root == NULL) {
    *root = node;
    return;
  }

  int rear = -1;
  int front = -1;
  Node *temp = {0};
  Node *queue[100];
  queue[++rear] = *root;

  while (front != rear) {
    temp = queue[++front];

    if (temp->left == NULL) {
      temp->left = node;
      return;
    } else {
      queue[++rear] = temp->left;
    }

    if (temp->right == NULL) {
      temp->right = node;
      return;
    } else {
      queue[++rear] = temp->right;
    }
  }
}

Node *get_deepest_right_node(Node *root) {
  Node *temp = {0};
  Node *queue[100];

  int rear = -1;
  int front = -1;
  queue[++rear] = root;

  while (front != rear) {
    temp = queue[++front];

    if (temp->left != NULL)
      queue[++rear] = temp->left;

    if (temp->right != NULL)
      queue[++rear] = temp->right;
  }

  return temp;
}

void del_deepest_right_node(Node *root, Node *d_node) {
  Node *temp;
  Node *queue[100];

  int rear = -1;
  int front = -1;
  queue[++rear] = root;

  while (front != rear) {
    temp = queue[++front];

    if (temp == d_node) {
      temp = NULL;
      free(d_node);
      return;
    }

    if (temp->right != NULL) {
      if (temp->right == d_node) {
        temp->right = NULL;
        free(d_node);
        return;
      } else {
        queue[++rear] = temp->right;
      }
    }

    if (temp->left != NULL) {
      if (temp->left == d_node) {
        temp->left = NULL;
        free(d_node);
        return;
      } else {
        queue[++rear] = temp->left;
      }
    }
  }
}

void delete_node(Node **root, int data) {
  if (*root == NULL) {
    printf("Tree is empty.\n");
    return;
  }

  if ((*root)->left == NULL && (*root)->right == NULL) {
    if ((*root)->data == data) {
      free(*root);
      *root = NULL;
      return;
    } else {
      printf("Node not found.\n");
      return;
    }
  }

  Node *temp;
  Node *queue[100];

  int rear = -1;
  int front = -1;
  queue[++rear] = *root;
  Node *key_node = NULL;

  while (front != rear) {
    temp = queue[++front];

    if (temp->data == data)
      key_node = temp;

    if (temp->left != NULL)
      queue[++rear] = temp->left;

    if (temp->right != NULL)
      queue[++rear] = temp->right;
  }

  if (key_node != NULL) {
    Node *deepest_node = get_deepest_right_node(*root);
    key_node->data = deepest_node->data;
    del_deepest_right_node(*root, deepest_node);
  } else {
    printf("Node not found.\n");
  }
}

Node *search_node(Node *root, int data) {
  if (root == NULL) return NULL;

  Node *temp;
  Node *queue[100];

  int rear = -1;
  int front = -1;
  queue[++rear] = root;

  while (front != rear) {
    temp = queue[++front];

    if (temp->data == data)
      return temp;

    if (temp->left != NULL)
      queue[++rear] = temp->left;

    if (temp->right != NULL)
      queue[++rear] = temp->right;
  }

  return NULL;
}

void inorder_traversal(Node *root) {
  if (root == NULL) return;
  inorder_traversal(root->left);
  printf("%d ", root->data);
  inorder_traversal(root->right);
}

int main() {
  Node *root = NULL;

  // Inserting nodes
  insert_node(&root, 20);
  insert_node(&root, 30);
  insert_node(&root, 40);
  insert_node(&root, 50);
  insert_node(&root, 60);
  insert_node(&root, 70);
  insert_node(&root, 80);

  // Inorder traversal
  printf("Inorder traversal of the given Binary Search Tree is: ");
  inorder_traversal(root);
  printf("\n");

  // Deleting a node
  int delete_value = 20;
  delete_node(&root, delete_value);
  printf("After deletion of %d: ", delete_value);
  inorder_traversal(root);
  printf("\n");

  // Inserting a new node
  int insert_value = 25;
  insert_node(&root, insert_value);
  printf("After insertion of %d: ", insert_value);
  inorder_traversal(root);
  printf("\n");

  // Searching for a node
  int target = 25;
  Node *search_result = search_node(root, target);
  if (search_result != NULL) {
    printf("Node %d found in the BST.\n", target);
  } else {
    printf("Node %d not found in the BST.\n", target);
  }

  return 0;
}

File: nim/ads/binary_tree.nim
type Node = ref object
  data: int
  left: Node
  right: Node

func newNode(data: int): Node =
  Node(data: data)

proc insert(root: var Node, data: int) =
  var node = newNode(data)
  var rear = -1
  var front = -1
  var temp = Node()
  var queue: array[100, Node]
  rear += 1
  queue[rear] = root

  while front != rear:
    front += 1
    temp = queue[front]

    if temp.left == nil:
      temp.left = node
      return

    rear += 1
    queue[rear] = temp.left

    if temp.right == nil:
      temp.right = node
      return

    rear += 1
    queue[rear] = temp.right

func deepestRightNode(root: Node): Node =
  var temp = Node()
  var queue: array[100, Node]

  var rear = -1
  var front = -1

  rear += 1
  queue[rear] = root

  while front != rear:
    front += 1
    temp = queue[front]

    if temp.left != nil:
      rear += 1
      queue[rear] = temp.left

    if temp.right != nil:
      rear += 1
      queue[rear] = temp.right

  return temp

proc delDeepestRightNode(root: Node, dNode: Node) =
  var temp: Node
  var queue: array[100, Node]

  var rear = -1
  var front = -1

  rear += 1
  queue[rear] = root

  while front != rear:
    front += 1
    temp = queue[front]

    if temp == dNode:
      temp = nil
      return

    if temp.right != nil:
      if temp.right == dNode:
        temp.right = nil
        return
      else:
        rear += 1
        queue[rear] = temp.right

    if temp.left != nil:
      if temp.left == dNode:
        temp.left = nil
        return
      else:
        rear += 1
        queue[rear] = temp.left

proc deleteNode(root: var Node, data: int) =
  if root.isNil():
    echo "Tree is empty"
    return
  
  if root.left.isNil() and root.right.isNil():
    if root.data == data:
      root = nil
      return
    else:
      echo "Node not found."
      return

  var temp: Node
  var queue: array[100, Node]

  var rear = -1
  var front = -1
  
  rear += 1
  queue[rear] = root
  var keyNode: Node

  while front != rear:
    front += 1
    temp = queue[front]

    if temp.data == data:
      keyNode = temp

    if temp.left != nil:
      rear += 1
      queue[rear] = temp.left

    if temp.right != nil:
      rear += 1
      queue[rear] = temp.right

    if keyNode != nil:
      let deepest = root.deepestRightNode()
      keyNode.data = deepest.data
      root.delDeepestRightNode(deepest)
    else:
      echo "Node not found."

func searchNode(root: Node, data: int): Node =
  if root == nil: return nil
  var temp = Node();
  var queue: array[100, Node]

  var rear = -1;
  var front = -1;

  rear += 1
  queue[rear] = root;

  while front != rear:
    front += 1
    temp = queue[front]

    if temp.data == data:
      return temp

    if temp.left != nil:
      rear += 1
      queue[rear] = temp.left

    if temp.right != nil:
      rear += 1
      queue[rear] = temp.right
  
  return nil

proc inorderTraversal(root: Node) =
  if root.isNil(): return
  inorderTraversal(root.left)
  stdout.write $root.data & " "
  inorderTraversal(root.right)

proc main() =
  var root = newNode(10)

  # Inseting
  root.insert(20)
  root.insert(30)
  root.insert(40)
  root.insert(50)

  # Inorder traversal
  echo "Inorder traversal of the given Binary Search Tree is: "
  root.inorderTraversal()
  echo()

  # Deleting a node
  var value = 50
  root.deleteNode(value)
  echo "After deletion of: " & $value
  root.inorderTraversal()
  echo()

  # Inserting a new node
  value = 60
  root.insert(value)
  echo "After insertion of: " & $value
  root.inorderTraversal()
  echo()

  # Searching for a node
  value = 10
  let search_result = root.searchNode(value)
  if search_result != nil:
    echo "Node " & $value & " found in the BST."
  else:
    echo "Node " & $value & " not found in the BST."

when isMainModule:
  main()
